nerfstudio/generative/deepfloyd.py

```diff
@@ -0,0 +1,207 @@
+
+    # load in model
+    def __init__(self, device: Union[torch.device, str]):
+        super().__init__()
+        self.device = device
+
+        self.text_encoder = T5EncoderModel.from_pretrained(
+            "DeepFloyd/IF-I-L-v1.0",
+            subfolder="text_encoder",
+            load_in_8bit=True,
+            variant="8bit",
+            device_map="auto",
+        )
+
+        self.pipe = IFPipeline.from_pretrained(
+            "DeepFloyd/IF-I-L-v1.0",
+            text_encoder=self.text_encoder,
+            safety_checker=None,
+            watermarker=None,
+            feature_extractor=None,
+            requires_safety_checker=False,
+            variant="fp16",
+            torch_dtype=torch.float16,
+        ).to(self.device)
+        
+        self.pipe.enable_attention_slicing(1)
+        self.pipe.unet.to(memory_format=torch.channels_last)
+
+        self.unet = self.pipe.unet
+        for p in self.unet.parameters():
+            p.requires_grad_(False)
+
+        self.scheduler = self.pipe.scheduler
+
+        self.num_train_timesteps = self.scheduler.config.num_train_timesteps
+        self.min_step = int(self.num_train_timesteps * 0.02)
+        self.max_step = int(self.num_train_timesteps * 0.98)
+
+
+        self.alphas: Float[Tensor, "..."] = self.scheduler.alphas_cumprod.to(
+            self.device
+        )
+
+    def delete_text_encoder(self): 
+        del self.text_encoder
+        del self.pipe
+        gc.collect()
+        torch.cuda.empty_cache()
+
+        self.pipe = IFPipeline.from_pretrained(
+            "DeepFloyd/IF-I-L-v1.0",
+            text_encoder=None,
+            safety_checker=None,
+            watermarker=None,
+            feature_extractor=None,
+            requires_safety_checker=False,
+            variant="fp16",
+            torch_dtype=torch.float16,
+        ).to(self.device)
+        
+        self.pipe.enable_attention_slicing(1)
+        self.pipe.unet.to(memory_format=torch.channels_last)
+
+        self.unet = self.pipe.unet
+        for p in self.unet.parameters():
+            p.requires_grad_(False)
+
+        self.scheduler = self.pipe.scheduler
+
+    def get_text_embeds(
+        self, prompt: Union[str, List[str]], negative_prompt: Union[str, List[str]]
+    ) -> TensorType[2, "max_length", "embed_dim"]: 
+        prompt = [prompt] if isinstance(prompt, str) else prompt
+        negative_prompt = [negative_prompt] if isinstance(negative_prompt, str) else negative_prompt
+        prompt_embeds, negative_embeds = self.pipe.encode_prompt(prompt, negative_prompt=negative_prompt)
+
+        return torch.cat([negative_embeds, prompt_embeds])
+
+    def sds_loss(
+        self,
+        text_embeddings, 
+        image, 
+        guidance_scale, 
+        grad_scaler,
+    ) -> torch.Tensor:
+        with torch.autocast(device_type="cuda", enabled=False):
+            image = F.interpolate(image.half(), (IMG_DIM, IMG_DIM), mode="bilinear", align_corners=False)
```

Comment:
Does `IMG_DIM` always need to be 64? If not, we might want to add it as an argument.

Reply:
We're only using the first model of Deepfloyd for SDS, which is trained to generate 64x64 images. Not sure what the results would look like if we used a different image size though. 

the other two models in the pipeline are the super resolution models that allow it to generate higher resolution images

Reply:
ok, feel free to ignore my comment